pkg/proxy/index/index_controller.go

```diff
@@ -19,441 +19,276 @@ package index
+		clientGetter: clientGetter,
+
+		clusterWorkspaceShardIndexer: clusterWorkspaceShardInformer.Informer().GetIndexer(),
+		clusterWorkspaceShardLister:  clusterWorkspaceShardInformer.Lister(),
+
+		shardClusterWorkspaceInformers: map[string]cache.SharedIndexInformer{},
+		shardClusterWorkspaceStopCh:    map[string]chan struct{}{},
+
+		workspaceShardNames: map[logicalcluster.Name]string{},
+		shardBaseURLs:       map[string]string{},
 	}
 
-	workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+	c.clusterWorkspaceHandler = cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			c.enqueueOrg(obj)
-			c.enqueueWorkspace(rootShard, tenancyv1alpha1.RootCluster, obj)
+			ws := obj.(*tenancyv1alpha1.ClusterWorkspace)
+			c.lock.RLock()
+			got := c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)]
+			c.lock.RUnlock()
+
+			if expected := ws.Status.Location.Current; got != expected {
+				c.lock.Lock()
+				defer c.lock.Unlock()
+				c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)] = expected
+			}
 		},
-		UpdateFunc: func(_, obj interface{}) {
-			c.enqueueOrg(obj)
-			c.enqueueWorkspace(rootShard, tenancyv1alpha1.RootCluster, obj)
+		UpdateFunc: func(old, obj interface{}) {
+			ws := obj.(*tenancyv1alpha1.ClusterWorkspace)
+
+			c.lock.RLock()
+			got := c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)]
+			c.lock.RUnlock()
+
+			if expected := ws.Status.Location.Current; got != expected {
+				c.lock.Lock()
+				defer c.lock.Unlock()
+				c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)] = expected
+			}
 		},
-		// TODO: handle deletes
-	})
-	if err := c.workspaceIndexer.AddIndexers(map[string]cache.IndexFunc{
-		currentShardIndex: func(obj interface{}) ([]string, error) {
-			if workspace, ok := obj.(*tenancyv1alpha1.ClusterWorkspace); ok {
-				return []string{workspace.Status.Location.Current}, nil
+		DeleteFunc: func(obj interface{}) {
+			if final, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				obj = final.Obj
 			}
-			return []string{}, nil
+			ws := obj.(*tenancyv1alpha1.ClusterWorkspace)
+
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			delete(c.workspaceShardNames, logicalcluster.From(ws).Join(ws.Name))
 		},
-	}); err != nil {
-		return nil, fmt.Errorf("failed to add indexer for ClusterWorkspace: %w", err)
 	}
 
-	workspaceShardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    func(obj interface{}) { c.enqueueOrgsForShard(obj) },
-		UpdateFunc: func(_, obj interface{}) { c.enqueueOrgsForShard(obj) },
+	clusterWorkspaceShardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			shard := obj.(*tenancyv1alpha1.ClusterWorkspaceShard)
+			c.lock.RLock()
+			got := c.shardBaseURLs[shard.Name]
+			c.lock.RUnlock()
+
+			if expected := shard.Spec.BaseURL; got != expected {
+				c.lock.Lock()
+				defer c.lock.Unlock()
+				c.shardBaseURLs[shard.Name] = expected
```

Comment:
Why not re-check now that we hold the lock? This is racy. If we can defer the check to the reconcile loop, why bother locking, could just enqueue always?